python-package/xgboost/distributed.py

```diff
@@ -0,0 +1,369 @@
+"""Dask extensions for distributed training. See xgboost/demo/dask for
+examples."""
+import platform
+from collections import defaultdict
+from toolz import first
+import numpy
+import pandas
+try:
+    import sparse
+    import scipy.sparse as ss
+except ImportError:
+    sparse = False
+    ss = False
+
+import logging
+from threading import Thread
+from .compat import distributed_get_worker
+import dask
+from dask.array import Array as DA
+from dask.dataframe import DataFrame as DDF
+from dask.dataframe import Series as DS
+try:
+    from dask.distributed import Client, get_client, wait
+except ImportError:
+    Client = None
+    get_client = None
+    wait = None
+try:
+    from dask.distributed import comm
+except ImportError:
+    comm = None
+from . import rabit
+from .core import DMatrix
+from .tracker import RabitTracker
+from dask import delayed
+
+
+def _start_tracker(host, n_workers):
+    """ Start Rabit tracker """
+    env = {'DMLC_NUM_WORKER': n_workers}
+    rabit_context = RabitTracker(hostIP=host, nslave=n_workers)
+    env.update(rabit_context.slave_envs())
+
+    rabit_context.start(n_workers)
+    thread = Thread(target=rabit_context.join)
+    thread.daemon = True
+    thread.start()
+    return env
+
+
+class RabitContext:
+    '''A context controling rabit initialization and finalization.'''
+    def __init__(self, args):
+        self.args = args
+
+    def __enter__(self):
+        rabit.init(self.args)
+        logging.debug('-------------- rabit say hello ------------------')
+
+    def __exit__(self, *args):
+        rabit.finalize()
+        logging.debug('--------------- rabit say bye ------------------')
+
+
+def concat(L):
```

Comment:
Instead of dispatching concatenation through `isinstance` checks, I recommend using `dask.dataframe.multi.concat`, here. This way Dask handles the types checks.

[Here's a link to the documentation](https://docs.dask.org/en/latest/dataframe-api.html?highlight=concat#dask.dataframe.multi.concat).

Reply:
@mt-jones Thanks for the pointer.

Reply:
@mt-jones I put `multi.concat` at the end of my type checkings since I still need to check for array instead of dataframe.  cudf should go into that route.  Also I got some weird hanging in `pytest` with `multi.concat` but can't reproduce outside of those fixtures used in tests.